iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -21,261 +21,490 @@ namespace iox
+        if (std::numeric_limits<uint64_t>::max() - additionalSeconds < m_seconds)
+        {
+            std::clog << __PRETTY_FUNCTION__
+                      << ": Applied values are out of range and would overflow, clamping to max value!" << std::endl;
+            m_seconds = std::numeric_limits<uint64_t>::max();
+            m_nanoseconds = NANOSECS_PER_SEC - 1U;
+        }
+        else
+        {
+            m_seconds += additionalSeconds;
+            m_nanoseconds = m_nanoseconds % NANOSECS_PER_SEC;
+        }
+    }
 }
 
-inline constexpr Duration::Duration(const struct timespec& value)
-    : Duration(static_cast<long double>(value.tv_sec) + static_cast<long double>(value.tv_nsec) / 1000000000.0)
+inline constexpr Duration::Duration(const struct timeval& value) noexcept
+    : Duration(static_cast<uint64_t>(value.tv_sec), static_cast<uint32_t>(value.tv_usec) * NANOSECS_PER_MICROSEC)
 {
 }
 
-inline constexpr Duration::Duration(const struct itimerspec& value)
-    : Duration(value.it_interval)
+inline constexpr Duration::Duration(const struct timespec& value) noexcept
+    : Duration(static_cast<uint64_t>(value.tv_sec), static_cast<uint32_t>(value.tv_nsec))
 {
 }
 
-inline constexpr Duration::Duration(const std::chrono::milliseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000.0)
+inline constexpr Duration::Duration(const struct itimerspec& value) noexcept
+    : Duration(value.it_interval)
 {
 }
 
-inline constexpr Duration::Duration(const std::chrono::nanoseconds& value)
-    : Duration(static_cast<long double>(value.count()) / 1000000000.0)
+inline constexpr Duration::Duration(const std::chrono::milliseconds& value) noexcept
 {
+    *this = Duration::milliseconds(value.count());
 }
 
-inline Duration& Duration::operator=(const std::chrono::milliseconds& right)
+inline constexpr Duration::Duration(const std::chrono::nanoseconds& value) noexcept
 {
-    durationInSeconds = static_cast<long double>(right.count()) / 1000.0;
-    return *this;
+    *this = Duration::nanoseconds(value.count());
 }
 
-inline constexpr Duration::Duration(const long double durationInSeconds)
-    : durationInSeconds(durationInSeconds >= 0.0 ? durationInSeconds : 0.0)
+inline Duration& Duration::operator=(const std::chrono::milliseconds& rhs) noexcept
 {
+    *this = Duration(rhs);
+    return *this;
 }
 
-template <typename T>
-inline constexpr T Duration::nanoSeconds() const
+inline constexpr uint64_t Duration::nanoSeconds() const noexcept
 {
-    return static_cast<T>(durationInSeconds * 1000000000.0);
+    constexpr uint64_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() / NANOSECS_PER_SEC};
+    constexpr uint64_t MAX_NANOSECONDS_BEFORE_OVERFLOW{std::numeric_limits<uint64_t>::max() % NANOSECS_PER_SEC};
+    constexpr Duration MAX_DURATION_BEFORE_OVERFLOW{MAX_SECONDS_BEFORE_OVERFLOW, MAX_NANOSECONDS_BEFORE_OVERFLOW};
+
+    if (*this > MAX_DURATION_BEFORE_OVERFLOW)
```

Comment:
Here I would maybe also write an internal free function to avoid code duplication.

```cpp
template<typename OnSuccessFunc>
uint64_t performConversion(const Duration & currentDuration, const Duration & maxDuration, const OnSuccessFunc & func) {
  if ( currentDuration > maxDuration ) {
        std::clog << __PRETTY_FUNCTION__ << ": Result of conversion would overflow, clamping to max value!"
                  << std::endl;
        return std::numeric_limits<uint64_t>::max();
  } 
  return OnSuccessFunc();
}

inline constexpr uint64_t Duration::nanoSeconds() const noexcept {
   // ...
   performConversion(*this, 
                      {MAX_SECONDS_BEFORE_OVERFLOW, MAX_NANOSECONDS_BEFORE_OVERFLOW}, 
                      [&]{ return m_seconds * NANOSECS_PER_SEC + m_nanoseconds; } );
}

```

Reply:
In this case I'd rather live with minimal code duplication